src/store.js

```diff
@@ -69,39 +48,58 @@ class Store {
     return this.store.getState().api.resources;
   }
 
-  all({ model } = {}) {
-    return Promise.resolve(Object.values(this.state[model.type]));
+  peek(id, { model } = {}) {
+    return toModel(model)(this.state[model.type][id]);
   }
 
-  where(query, { remote = true, model } = {}) {
-    const local = queryData(Object.values(this.state[model.type]), query);
-    if (local.length > 0 || remote === false) {
-      return Promise.resolve(local);
-    }
-
-    return this.dispatch(remoteRead({ resource: model, query }))
-      .then(() => this.where(query, { remote: false, model }));
+  /**
+   * @param {String|Number} id
+   * @param {Object} options
+   * @param {Model} options.model
+   * @returns {RecordPromise<Model|null>}
+   */
+  find(id, { model } = {}) {
+    const endpoint = `${model.endpoint('read')}/${id}`;
+    const promise = this.dispatch(apiRead({ resource: model, endpoint }));
+    const mapper = toModel(model);
+    const selector = createSelector(
+      (state) => state.api.resources[model.type],
+      (resources = {}) => mapper(resources[id]),
+    );
+
+    return createRecordPromise({ promise, selector, store: this.store });
   }
 
-  fromIndex({ id, type }) {
-    return this.state[type][id];
+  peekAll({ model } = {}) {
+    return mapToModel(model)(Object.values(this.state[model.type]));
   }
 
-  findById(id, { remote = true, model } = {}) {
-    const local = this.fromIndex({ id, type: model.type });
-    if (local || remote === false) {
-      return Promise.resolve(local);
-    }
+  findAll({ model } = {}) {
+    const promise = this.dispatch(apiRead({ resource: model }));
+    const mapper = mapToModel(model);
+    const selector = createSelector(
+      (state) => state.api.resources[model.type],
+      (resources = {}) => mapper(Object.values(resources)),
+    );
 
-    const endpoint = `${this.endpoint('read')}/${id}`;
-
-    return this.dispatch(remoteRead({ resource: model, endpoint }))
-      .then(() => this.findById(id, { remote: false, model }));
+    return createRecordPromise({ promise, selector, store: this.store });
   }
 
-  find({ id, ...query }, { remote = true, model } = {}) {
-    if (id) return this.findById(id, { remote, model });
-    return this.where(query, { remote, model }).then((results) => results[0]);
+  query(query, { model } = {}) {
+    const endpoint = queryEndpoint(model.endpoint('read'), query);
+    const promise = this.dispatch(apiRead({ resource: model, endpoint })).then(({ data }) => data);
+    this.dispatch(apiIndexAsync({ hash: endpoint, promise }));
+    const selector = createSelector(
+      (state) => state.api.indices[endpoint],
+      (state) => state.api.resources[model.type],
+      (index = []) => {
+        const results = index.map((item) => this.peek(item.id, { model }));
+        results.isFetching = index.isFetching;
+        return results;
+      }
+    );
+
+    return createRecordPromise({ promise, selector, store: this.store });
```

Comment:
This abstraction feels fairly clean now, I think. It felt pretty forced earlier, but now I feel like things live in the right places. But would love feedback on this.